examples/all-clusters-app/all-clusters-common/shell_extension/control.cpp

```diff
@@ -0,0 +1,260 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "control.hpp"
+
+#include "app/common/gen/enums.h"
+#include "app/util/DataModelHandler.h"
+#include "app/util/basic-types.h"
+#include "core/CHIPError.h"
+#include "core/NodeId.h"
+#include "messaging/ExchangeContext.h"
+#include "messaging/ExchangeDelegate.h"
+#include "protocols/secure_channel/SessionEstablishmentDelegate.h"
+#include "support/logging/CHIPLogging.h"
+#include "transport/FabricTable.h"
+#include "transport/SessionHandle.h"
+#include "transport/raw/PeerAddress.h"
+#include <app/server/Server.h>
+#include <bits/stdint-uintn.h>
+#include <controller/CHIPDevice.h>
+#include <controller/InteractionPeer.h>
+#include <credentials/CHIPCert.h>
+#include <cstdio>
+#include <lib/shell/commands/Help.h>
+#include <lib/shell/streamer.h>
+#include <lib/support/BytesToHex.h>
+#include <lib/support/CodeUtils.h>
+#include <memory>
+#include <protocols/secure_channel/CASESession.h>
+#include <zap-generated/CHIPClusters.h>
+
+using chip::CASESession;
+using chip::FabricIndex;
+using chip::NodeId;
+using chip::Optional;
+using chip::SecureSessionMgr;
+using chip::Server;
+using chip::SessionEstablishmentDelegate;
+using chip::SessionHandle;
+using chip::Controller::InteractionPeer;
+using chip::Controller::InteractionPeerDelegate;
+using chip::Messaging::ExchangeContext;
+using chip::Messaging::ExchangeDelegate;
+using chip::Shell::Engine;
+using chip::Shell::PrintCommandHelp;
+using chip::Shell::shell_command_t;
+using chip::Shell::streamer_get;
+using chip::Shell::streamer_printf;
+using chip::Transport::FabricInfo;
+using chip::Transport::PeerAddress;
+
+namespace {
+
+class Device : public InteractionPeerDelegate, public SessionEstablishmentDelegate, public ExchangeDelegate
```

Comment:
I've tried to used `chip::Controller::Device`. This class has a built in state machine and requires a PASE session or a non-volatile state storage of paired devices to drive it. Neither is available on an accessory. Directly calling `EstablishConnectivity` will return `ERROR_INCORRECT_STATE`. I've added a function in `Device` class to explicitly set the state and `EstablishConnectivity` can be used to establish session and send cluster functions. That's in fact how https://github.com/project-chip/connectedhomeip/pull/9103 was originally implemented. @msandstedt Would you provide some code snippets on how you're using the `Device` outside of `DeviceController`?

In the discussion https://github.com/project-chip/connectedhomeip/pull/9103#discussion_r691751137 we found such work around to be hacky. To decouple PASE session we added the `InteractionPeerDelegate` and the new `Device` class. Another choice is to totally rewrite the `Device` class and the controller which will be a huge project. My suggestion is to unblock device-to-device functionalities first so that we can go on with binding and OTA. At the same time we can review and plan the refactor of the clustger clients.

Reply:
> I've tried to used `chip::Controller::Device`. This class has a built in state machine and requires a PASE session or a non-volatile state storage of paired devices to drive it. Neither is available on an accessory. Directly calling `EstablishConnectivity` will return `ERROR_INCORRECT_STATE`. I've added a function in `Device` class to explicitly set the state and `EstablishConnectivity` can be used to establish session and send cluster functions. That's in fact how #9103 was originally implemented. @msandstedt Would you provide some code snippets on how you're using the `Device` outside of `DeviceController`?

Yes, this is easily fixed with the following:

```
diff --git a/src/controller/CHIPDevice.cpp b/src/controller/CHIPDevice.cpp
index c41472957..5d11c5b63 100644
--- a/src/controller/CHIPDevice.cpp
+++ b/src/controller/CHIPDevice.cpp
@@ -533,6 +533,7 @@ void Device::OperationalCertProvisioned()
 
     Persist();
     CloseSession();
+    mState = ConnectionState::NotConnected;
 }
 
```

Reply:
> In the discussion #9103 (comment) we found such work around to be hacky. To decouple PASE session we added the InteractionPeerDelegate and the new Device class. Another choice is to totally rewrite the Device class and the controller which will be a huge project. My suggestion is to unblock device-to-device functionalities first so that we can go on with binding and OTA. At the same time we can review and plan the refactor of the clustger clients.

We struggled with the Device class too.  But in fact, I found the needed workarounds to be few.  The diff above is the only thing that I know of that must be changed in the sdk to do this.  `chip::Controller::Device` does not have nearly as much state baggage as `chip::Controller::DeviceController` does.

I really think using the existing Device class should be considered.  I believe the sdk is closer to allowing for such a thing than people might believe.  We have been successfully using the Device object completely outside of the controller since July 8th.

Reply:
Indeed, with the suggested change or setting up a mutator to update device state before calling `EstablishConnectivity`, it should be fairly straightforward to extend/use existing `chip::Controller:Device`. I understand that it's a hack to reach `Device::WarmupCASESession()` but it supports using existing bootstraps we have there than defining everything again in `Connect()` here.
The callstack in `ControlConnectHandler` therefore very much reduces to:
```cpp
mDevice.Init(...) // chip::Controller::Device::Init(...)
mDevice.SetActive(true);
mDevice.OperationalCertProvisioned();
mDevice.EstablishConnectivity();
```